### DISCUSSION TITLE
mod_muc_log alternative for logging to DB

### DISCUSSION THREAD

**Discussion by Dunedan** [processone/ejabberd#4191] [Questions] [answered]
Is there an alternative to `mod_muc_log` somewhere which supports logging to a relational database instead of files?

**Comment by Dunedan** [comment]
To follow up on that: After looking at the code I realized `mod_muc` explicitly calls into functions of `mod_muc_log` for logging. If I'm not mistaken that makes it impossible to write a module doing logging in another way, without modifying `mod_muc` as well.

I'm not familiar with Erlang, but I wonder if there would be a possibility to handle that differently, so alternative MUC logging modules could exist.

**Comment by badlop** [comment]
I am not aware of any other module for MUC room logging.

Right, as you noticed, right now mod_muc_room directly calls erlang functions in mod_muc_log. This could be replaced with calls to [ejabberd_hooks](https://docs.ejabberd.im/developer/guide/#hooks), and that would allow developers to write alternative modules, and administrators to enable the desired one/ones.

As a first experiment, I've applied that proposed change in my fork of ejabberd, see https://github.com/badlop/ejabberd/commit/52629756628ce4033644383e2cda78cc4e932688

There's also an example module, see https://github.com/badlop/ejabberd/commit/0080c8e4f11e7fea19686862077b488ddd5859fc

**Comment by Dunedan** [comment]
I played around with that a bit and implemented a rudimentary module for logging to a database. Unfortunately, without proper Erlang experience I doubt I'll be able to get that into a polished state for the time being.

Your changes to introduce ejabberd hooks do work great as far as I can see, so I'd appreciate if you could merge that to offer flexibility for additional logging modules.

**Reply by badlop** [reply, marked as answer]
Ok, I've merged that patch into ejabberd: https://github.com/processone/ejabberd/commit/d04f92d44e532e6ef60f896ca7ae9a80d3ccd709
